File: Swift/BOJ/20000/23304.swift
//
//  23304.swift
//  Algorithm
//
//  Created by 유정주 on 2023/06/16.
//

import Foundation

func isAkaraka(_ string: [Character]) -> Bool {
    if string.count == 1 {
        return true
    }
    
    let length = string.count
    let mid = Int(floor(Double(length) / 2.0))
    
    let front = Array(string[..<mid])
    var tail: [Character] = []
    if length % 2 == 0 {
        tail = Array(string[mid...])
    } else {
        tail = Array(string[(mid+1)...])
    }
    
    if front != tail.reversed() {
        return false
    }
    
    return isAkaraka(front)
}


let str = Array(readLine()!)

let result = isAkaraka(str)

if result {
    print("AKARAKA")
} else {
    print("IPSELENTI")
}

File: Python/BOJ/문자열/23304.py
//
//  23304.py
//  Algorithm
//
//  Created by 유정주 on 2023/06/16.
//

import sys; readLine = sys.stdin.readline

def isAkaraka(string):
    if len(string) == 1:
        return True
        
    length = len(string)
    mid = length // 2

    front = string[0:mid]
    if length % 2 == 0:
        tail = string[mid:]
    else:
        tail = string[mid+1:]

    if front != tail[::-1]:
        return False
    
    return isAkaraka(front)

str = readLine().rstrip()

result = isAkaraka(str)
if result:
    print("AKARAKA")
else:
    print("IPSELENTI")
